BJS2-96233

Monkeylust opened the pull request:
BJS2-96233
// In progress

Monkeylust committed:
BJS2-96233
// In progress

Monkeylust committed:
BJS2-96233
// In progress

Monkeylust committed:
BJS2-96233
// In progress

Monkeylust committed:
BJS2-96233
// In progress

Monkeylust committed:
BJS2-96233
// In progress

Monkeylust committed:
BJS2-96233
// In progress

Monkeylust committed:
BJS2-96233
// In progress

Monkeylust committed:
BJS2-96233
// In progress

mipkasnow commented on src/main/java/school/faang/user_service/controller/career/CareerController.java:
```diff
@@ -0,0 +1,49 @@
+
+@RestController
+@RequestMapping("/career")
+@RequiredArgsConstructor
```
"Аннотации в порядке важности" (с) Дима
@RequiredArgsConstructor
@RequestMapping("/career")
@RestController

mipkasnow commented on src/main/java/school/faang/user_service/controller/career/CareerController.java:
```diff
@@ -0,0 +1,49 @@
+    public ResponseEntity<CareerDto> updateCareer(@PathVariable long careerId,
+                                                  @Valid @RequestBody UpdateCareerDto updateCareerDto) {
+        long userId = userContext.getUserId();
+        CareerDto updatedCareer = careerService.updateCareer(userId, careerId, updateCareerDto);
```
long userId = userContext.getUserId();
Лучше единообразно внутри сервиса вызывать этот метод, тогда и третий параметр в методе updateCareer не нужен

mipkasnow commented on src/main/java/school/faang/user_service/dto/career/CareerDateDto.java:
```diff
@@ -0,0 +1,9 @@
+
+import java.time.LocalDate;
+
+public interface CareerDateDto {
```
Интерфейс? Должен быть рекорд же

mipkasnow commented on src/main/java/school/faang/user_service/dto/career/CareerDto.java:
```diff
@@ -0,0 +1,16 @@
+import java.time.LocalDate;
+
+@Builder
+@JsonInclude(JsonInclude.Include.NON_NULL)
```
Подлей мастер к себе ветку - там это настройка глобально теперь ставится в маппере => аннотация не нужна

mipkasnow commented on src/main/java/school/faang/user_service/dto/career/CreateCareerDto.java:
```diff
@@ -0,0 +1,11 @@
+public record CreateCareerDto(@NotNull LocalDate from,
+                              LocalDate to,
+                              @NotBlank String company,
+                              @NotBlank String position) implements CareerDateDto {
```
implements CareerDateDto - чет это лишнее по моему)

mipkasnow commented on src/main/java/school/faang/user_service/dto/career/UpdateCareerDto.java:
```diff
@@ -0,0 +1,15 @@
+import java.time.LocalDate;
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public record UpdateCareerDto(@NotNull LocalDate from,
```
"Надо дто назвать в едином формате" (с) Дима
CareerUpdateDto
CareerDto
CareerCreateDto

mipkasnow commented on src/main/java/school/faang/user_service/service/career/CareerService.java:
```diff
@@ -0,0 +1,58 @@
+import school.faang.user_service.validator.career.CareerValidator;
+
+@Service
+@RequiredArgsConstructor
```
Аннотации в порядке важности

mipkasnow commented on src/main/java/school/faang/user_service/service/career/CareerService.java:
```diff
@@ -0,0 +1,58 @@
+        careerRepository.delete(career);
+    }
+
+    public CareerDto updateCareer(long userId, long careerId, UpdateCareerDto updateCareerDto) {
```
long userId можно же внутри метода вызывать

mipkasnow reviewed (commented):
В целом норм, оставил пару комментариев.
Еще момент - коммиты надо осмысленно называть - что сделано/исправлено

Monkeylust committed:
BJS2-96233
// In progress

Monkeylust committed:
Merge branch 'refs/heads/hippogriff-master-stream12' into hippogriff12/Monkeylust/BJS2-96233-Career

# Conflicts:
#	src/main/java/school/faang/user_service/UserServiceApplication.java

Monkeylust committed:
BJS2-96233
// Fixed code based on Mikhail's comments

Monkeylust committed:
BJS2-96233
// Fixed checkstyle

Monkeylust committed:
BJS2-96233
// Fixed checkstyle

20cognosce commented on src/main/java/school/faang/user_service/controller/career/CareerController.java:
```diff
@@ -0,0 +1,48 @@
+@RestController
+public class CareerController {
+    private final CareerService careerService;
+    private final UserContext userContext;
```
Зачем он тут тебе

20cognosce commented on src/main/java/school/faang/user_service/dto/career/CareerDto.java:
```diff
@@ -0,0 +1,14 @@
+import java.time.LocalDate;
+
+@Builder
+public record CareerDto(long id,
```
а почему тут не унаследовался от BaseCareerDtoWithDates

Monkeylust replied:
Потому, что мне изначально об этом не сказали и у меня крашится проект в CareerMapper из-за наследования, а как исправить я не знаю.

Monkeylust replied:
<img width="1920" height="1032" alt="image" src="https://github.com/user-attachments/assets/9c23e3dd-5dc7-4bb5-8234-9b3f5ed34f31" />

20cognosce commented on src/main/java/school/faang/user_service/dto/career/CareerUpdateDto.java:
```diff
@@ -0,0 +1,21 @@
+    @NotBlank
+    private final String company;
+    @NotBlank
+    private final String position;
```
@NotBlank не здесь, а в конструкторе нужно проставить
Нужно протестировать, что они работают

20cognosce commented on src/main/java/school/faang/user_service/dto/career/CreateCareerDto.java:
```diff
@@ -0,0 +1,21 @@
+    @NotBlank
+    private final String company;
+    @NotBlank
+    private final String position;
```
Аналогично

20cognosce commented on src/main/java/school/faang/user_service/service/career/CareerService.java:
```diff
@@ -0,0 +1,59 @@
+        Career career = careerMapper.toCareer(createCareerDto);
+        career.setUser(user);
+        career = careerRepository.save(career);
+        return careerMapper.toCareerDto(career);
```
Хорошо, но добавляй пустые строки между логическими этапами, чего всё слиплось

```
CareerValidator.validateCareerDates(createCareerDto);

long requesterId = userContext.getUserId();
User user = userRepository.getByIdOrThrow(requesterId);

Career career = careerMapper.toCareer(createCareerDto);
career.setUser(user);

career = careerRepository.save(career);
return careerMapper.toCareerDto(career);
```

20cognosce commented on src/main/java/school/faang/user_service/service/career/CareerService.java:
```diff
@@ -0,0 +1,59 @@
+        long requesterId = userContext.getUserId();
+        User user = userRepository.getByIdOrThrow(requesterId);
+        Career career = careerRepository.getByIdOrThrow(careerId);
+        CareerValidator.validateOwner(career, user.getId());
```
А нужно ли тут проверять владельца? Чужую карьеру что ли просмотреть нельзя? Ну это странно, смотреть наверно стоит разрешить

20cognosce reviewed (changes requested):
Хорошо, доделать немного и начинай писать тесты

Monkeylust committed:
BJS2-96233
// 1) Убрал лишнее поле в CareerController
2) Исправил аннотации в CreateCareerDto и UpdateCareerDto
3) Изменил имя на UpdateCareerDto

Monkeylust committed:
BJS2-96233
// 1) Удалил валидацию владельца поста при запросе на просмотр.

Monkeylust committed:
BJS2-96233
// Переделал CareerDto, унаследовал от BaseCareerDtoWithDates

20cognosce commented on src/main/java/school/faang/user_service/dto/career/CareerDto.java:
```diff
@@ -0,0 +1,28 @@
+    private final String company;
+    private final String position;
+
+    @Builder
```
Лучше на класс вешать и указывать в нем параметр callSuper = true
или есть @SuperBuilder аннотация вроде бы

чтобы и поля родителя в него попадали

20cognosce reviewed (approved):
Принято!
